### PR TITLE
Includes new methods to subscribe and unsubscribe for Cprsubscription

### PR DESCRIPTION
**Pull request by NakulSeerpu**

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,3 +1,6 @@
+1.2.0
+* Updated CprSubscription to subscribe and unsubscribe for CPR events.
```
Comment (Araballi) [inline CHANGES.md:2]: Change version not give details of changes in this version, update version details as discussed

File: CHANGES.md [hunk 2]
```diff
@@ -1,3 +1,11 @@
+1.2.0
+* Updated CprSubscription to subscribe and unsubscribe for CPR events,it includes these methods
+ - Subscribe To The Events By CPR number
+ - Subscribe To The Events By PersonId
+ - Unsubscribe By CPR number
+ - Unsubscribe By PersonId
+ -Get Events
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply CHANGES.md]: Got it

File: README.md [hunk 3]
```diff
@@ -107,6 +107,25 @@ If not one of the well-described tests, the Fake Provider exhibits the following
 - If the CPR number ends in `000`, `001` or `002` it returns NotFound
 - The Fake provider returns random data, using the CPR number as the seed. This ensures recurring calls return the same response
 
+## CPR Subscription
+
+When ever there is a change in Data of Citizen, The respsective provider will raise events.
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply README.md]: Changed as suggested

File: README.md [hunk 4]
```diff
@@ -107,6 +107,25 @@ If not one of the well-described tests, the Fake Provider exhibits the following
...
+ 
+Example -
+ [Datafordeler (Events)](https://confluence.datafordeler.dk/pages/viewpage.action?pageId=17137834#H%C3%A6ndelserp%C3%A5Datafordeleren-Brugsscenarier)
+ 
+- Subscribe To The Events By CPR number
+In Order to recieve events, you must subscribe with the `SubscriptionId` and `ConfigurationId` and the `CPR number` for which the events needs to be received.
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply README.md]: Changed as suggested

File: sample/Kmd.Logic.Cpr.Client.Sample/AppConfiguration.cs [hunk 5]
```diff
@@ -9,5 +10,15 @@ internal class AppConfiguration
         public LogicTokenProviderOptions TokenProvider { get; set; } = new LogicTokenProviderOptions();
 
         public CprOptions Cpr { get; set; } = new CprOptions();
+
+        public Guid CprPersonId { get; set; } = Guid.Parse("5e3d9df6-d082-467a-88bd-dca56edc7328");
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/AppConfiguration.cs:14]: We would have , but now the return type changed to bool.

File: README.md [hunk 6]
```diff
@@ -107,6 +107,25 @@ If not one of the well-described tests, the Fake Provider exhibits the following
...
+ 
+- Subscribe To The Events By PersonId
+To recieve events, you must subscribe with the `SubscriptionId` and `ConfigurationId` and the `CPR Person ID` for which the events needs to be received.
+ 
+- Unsubscribe By CPR number
+To Stop receiveing the events, you must unsubscribe the previously created subscription either with `CPR number` or `CPR Person ID`.
+ 
+- Get Events
+To generate these events for which you have subscibed, use `Get Events` along with your `SubscriptionId` and `ConfigurationId` along with the desired time period.
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply README.md]: Changed as suggested

File: src/Kmd.Logic.Cpr.Client/CprClient.cs [hunk 7]
```diff
@@ -87,7 +89,7 @@ public async Task<Citizen> GetCitizenByCprAsync(string cpr)
         /// <exception cref="SerializationException">Unable process the service response.</exception>
         /// <exception cref="LogicTokenProviderException">Unable to issue an authorization token.</exception>
         /// <exception cref="CprConfigurationException">Invalid CPR configuration details.</exception>
-        public async Task<Citizen> GetCitizenByIdAsync(Guid id)
+        public async Task<object> GetCitizenByIdAsync(Guid id)
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply src/Kmd.Logic.Cpr.Client/CprClient.cs]: Reverted back.

File: src/Kmd.Logic.Cpr.Client/CprClient.cs [hunk 8]
```diff
@@ -123,6 +125,145 @@ public async Task<IList<CprProviderConfigurationModel>> GetAllCprConfigurationsA
...
         }
 
+         /// <summary>
+        /// Subscribes for CPR events by CPR number.
+        /// </summary>
+        /// <param name="cpr">The CPR number.</param>
+        /// <returns>The Saved CprPersonId.</returns>
+        /// <exception cref="ValidationException"> When subscriptionId or CPR number is null.</exception>
+        /// <exception cref="SerializationException">Unable process the service response.</exception>
+        public async Task<Guid> SubscribeByCprAsync(string cpr)
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply src/Kmd.Logic.Cpr.Client/CprClient.cs]: Changed to bool as suggested.

File: src/Kmd.Logic.Cpr.Client/CprClient.cs [hunk 9]
```diff
@@ -123,6 +125,145 @@ public async Task<IList<CprProviderConfigurationModel>> GetAllCprConfigurationsA
...
+        public async Task<bool> UnsubscribeByCprAsync(string cpr)
+        {
+            var client = this.CreateClient();
+
+            using (var response = await client.UnsubscribeByCprWithHttpMessagesAsync(
+                 subscriptionId: this.options.SubscriptionId,
+                 cpr: cpr,
+                 configurationId: this.options.CprConfigurationId).ConfigureAwait(false))
+             {
+                switch (response.Response.StatusCode)
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply src/Kmd.Logic.Cpr.Client/CprClient.cs]: Changed as suggested.

File: src/Kmd.Logic.Cpr.Client/CprClient.cs [hunk 10]
```diff
@@ -123,6 +125,145 @@ public async Task<IList<CprProviderConfigurationModel>> GetAllCprConfigurationsA
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets citizen events for the nominated period.
+        /// </summary>
+        /// <param name="dateFom">Date from which events are required.</param>
+        /// <param name="dateTo">Date till which events are required.</param>
+        /// <param name="pageNo">Page number.</param>
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply src/Kmd.Logic.Cpr.Client/CprClient.cs]: Changed as suggested.

File: src/Kmd.Logic.Cpr.Client/CprClient.cs [hunk 11]
```diff
@@ -123,6 +125,145 @@ public async Task<IList<CprProviderConfigurationModel>> GetAllCprConfigurationsA
...
+        public async Task<Guid> SubscribeByIdAsync(Guid id)
+        {
+             var client = this.CreateClient();
+
+             using (var response = await client.SubscribeByIdWithHttpMessagesAsync(
+                subscriptionId: this.options.SubscriptionId,
+                id: id,
+                request: this.CprSubscription).ConfigureAwait(false))
+            {
+                switch (response.Response.StatusCode)
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply src/Kmd.Logic.Cpr.Client/CprClient.cs]: changed as suggested

File: sample/Kmd.Logic.Cpr.Client.Sample/AppConfiguration.cs [hunk 12]
```diff
@@ -9,5 +10,15 @@ internal class AppConfiguration
...
+
+        public DateTime DateFrom { get; set; } = DateTime.Today.AddMonths(-2);
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/AppConfiguration.cs]: Removed

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 13]
```diff
@@ -59,8 +59,16 @@ private static async Task Run(AppConfiguration configuration)
             using (var httpClient = new HttpClient())
             using (var tokenProviderFactory = new LogicTokenProviderFactory(configuration.TokenProvider))
             {
+                Console.WriteLine("Please make your choice, choose '1' to test regarding CprConfiguration and '2' to test regarding CprSubscription");
```
[the remark this answers is not in the document]
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: Removed

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 14]
```diff
@@ -50,6 +50,12 @@ private static void InitLogger()
 
         private static async Task Run(AppConfiguration configuration)
         {
+            var cprPersonId = Guid.Parse("5e3d9df6-d082-467a-88bd-dca56edc7328");
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:53]: No need for this. Use the Id off the citizen object

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 15]
```diff
@@ -50,6 +50,12 @@ private static void InitLogger()
...
+            var datefrom = DateTime.Today.AddMonths(-2);
+            var dateTo = DateTime.Today;
+            var pageNo = 1;
+            var pageSize = 1;
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:57]: Maybe set pageSize to 10 or 100?

File: README.md [hunk 16]
```diff
@@ -107,6 +107,22 @@ If not one of the well-described tests, the Fake Provider exhibits the following
...
 - The Fake provider returns random data, using the CPR number as the seed. This ensures recurring calls return the same response
 
+## CPR Subscription
+
+Whenever there is a change in information about a Citizen the CPR provider will raise events.
+ 
+Example -
+ [Datafordeler (Events)](https://confluence.datafordeler.dk/pages/viewpage.action?pageId=17137834#H%C3%A6ndelserp%C3%A5Datafordeleren-Brugsscenarier)
+ 
+- Subscribe To The Events By CPR number or by PersonId
```
Comment (base2browns) [inline README.md:117]: Subscribe to the events by CPR Number or by Person Id

File: README.md [hunk 17]
```diff
@@ -107,6 +107,22 @@ If not one of the well-described tests, the Fake Provider exhibits the following
...
+In Order to recieve events, you must subscribe with the `SubscriptionId` and `ConfigurationId` and the `CPR number` or by `CPR Person ID`  for which the events needs to be received.
```
Comment (base2browns) [inline README.md:118]: In order to receive events you must subscribe using the `CPR number` or `CPR Person ID` of the individual.

File: README.md [hunk 18]
```diff
@@ -107,6 +107,22 @@ If not one of the well-described tests, the Fake Provider exhibits the following
...
+ 
+- Unsubscribe By CPR number or by PersonId
+To Stop receiveing the events, you must unsubscribe the previously created subscription either with `CPR number` or `CPR Person ID`.
```
Comment (base2browns) [inline README.md:121]: To stop receiving the events you must unsubscribe the previously created subscription using the `CPR number` or `CPR Person ID` of the individual.

File: README.md [hunk 19]
```diff
@@ -107,6 +107,22 @@ If not one of the well-described tests, the Fake Provider exhibits the following
...
+ 
+- Get Events
+To fetch the events for which you have subscribed, use Get Events along with your SubscriptionId and ConfigurationId and the desired time period.
```
Comment (base2browns) [inline README.md:124]: The unfiltered GetEvents method still exists doesn't it? If so, shouldn't we refer to something like GetFilteredEvents?

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 15 above)
Comment (base2browns) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:57]: Do we need these settings here at all? Can't we just put them directly in the request?

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 20]
```diff
@@ -96,6 +102,45 @@ private static async Task Run(AppConfiguration configuration)
                 var citizen = await cprClient.GetCitizenByCprAsync(configuration.CprNumber).ConfigureAwait(false);
 
                 Log.Information("Citizen data: {@Citizen}", citizen);
+
+                var isSubscribeByCprSuccess = await cprClient.SubscribeByCprAsync(configuration.CprNumber).ConfigureAwait(false);
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:106]: isSubscribeByCprSuccess is quite long winded. Can't we just use `success` and then reuse the variable?

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 21]
```diff
@@ -96,6 +102,45 @@ private static async Task Run(AppConfiguration configuration)
...
+
+                if (!isSubscribeByCprSuccess)
+                {
+                    Log.Error("Invalid CPR Number {@CprNumber}", configuration.CprNumber);
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:110]: Should we stop here then?

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 22]
```diff
@@ -96,6 +102,45 @@ private static async Task Run(AppConfiguration configuration)
...
+                }
+
+                Log.Information("Subscribed successfully for CprNumber {CprNumber}", configuration.CprNumber);
+
+                var isSubscribeByIdSuccess = await cprClient.SubscribeByIdAsync(cprPersonId).ConfigureAwait(false);
+
+                if (!isSubscribeByIdSuccess)
+                {
+                    Log.Error("Invalid CPR PersonId {personId}", cprPersonId);
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:119]: Should we stop here then?

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 23]
```diff
@@ -96,6 +102,45 @@ private static async Task Run(AppConfiguration configuration)
...
+                }
+
+                var isUnsubscribeByIdSuccessful = await cprClient.UnsubscribeByIdAsync(cprPersonId).ConfigureAwait(false);
+
+                if (isUnsubscribeByIdSuccessful)
+                {
+                    Log.Information("Unsubscribed successfully for personId {personId}", cprPersonId);
+                }
+
+                var citizenList = await cprClient.GetAllCprEvents(datefrom, dateTo, pageNo, pageSize).ConfigureAwait(false);
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:138]: This is the unfiltered query. I welcome you calling it but I would put it before all the subscribe/unsubscribe logic

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 20 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: Changed as suggested

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 21 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: returning now

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 14 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: removed and making use of citizen object

File: README.md (hunk 16 above)
Comment (NakulSeerpu) [reply README.md]: changed as suggested

File: README.md (hunk 18 above)
Comment (NakulSeerpu) [reply README.md]: changed as suggested

File: README.md (hunk 19 above)
Comment (NakulSeerpu) [reply README.md]: Changed method name to GetFilteredEvents

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 22 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: yes, returning now.

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 23 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: Added this method at the beginning .

File: README.md (hunk 17 above)
Comment (NakulSeerpu) [reply README.md]: changed as suggested

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 15 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: Passing those as parameters of method directly

**Comment by Araballi** [comment]
LGTM

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs [hunk 24]
```diff
@@ -96,6 +96,48 @@ private static async Task Run(AppConfiguration configuration)
...
+
+                var citizenList = await cprClient.GetAllCprEvents(DateTime.Today.AddMonths(-2), DateTime.Today, 1, 10).ConfigureAwait(false);
+
+                if (citizenList == null)
+                {
+                    Log.Error("Error in retriving citizen list");
+                    return;
+                }
+
+                var isSubscribeSuccess = await cprClient.SubscribeByCprAsync(configuration.CprNumber).ConfigureAwait(false);
```
Comment (base2browns) [inline sample/Kmd.Logic.Cpr.Client.Sample/Program.cs:108]: can't we just use success?

File: src/Kmd.Logic.Cpr.Client/CprClient.cs [hunk 25]
```diff
@@ -123,6 +123,105 @@ public async Task<IList<CprProviderConfigurationModel>> GetAllCprConfigurationsA
...
+                return response.Response.IsSuccessStatusCode;
+            }
+        }
+
+        /// <summary>
+        /// Gets citizen events for the nominated period.
+        /// </summary>
+        /// <param name="dateFom">Query events from this date and time.</param>
+        /// <param name="dateTo">Query events to this date and time.</param>
+        /// <param name="pageNo">The page number to query.</param>
```
Comment (base2browns) [inline src/Kmd.Logic.Cpr.Client/CprClient.cs:209]: The page number to query, starting at 1.

File: README.md [hunk 26]
```diff
@@ -107,6 +107,22 @@ If not one of the well-described tests, the Fake Provider exhibits the following
 - If the CPR number ends in `000`, `001` or `002` it returns NotFound
 - The Fake provider returns random data, using the CPR number as the seed. This ensures recurring calls return the same response
 
+## CPR Subscription
+
+Whenever there is a change in information about a Citizen the CPR provider will raise events.
```
Comment (base2browns) [inline README.md:112]: We should probably clarify here:

The events raised will include citizens you have no records for. You can either:
a. Call `GetAllCprEvents` and determine the relevant events yourself
b. Subscribe for events about specific citizens then call `GetSubscribedCprEvents` to return just those you are interested in

File: sample/Kmd.Logic.Cpr.Client.Sample/Program.cs (hunk 24 above)
Comment (NakulSeerpu) [reply sample/Kmd.Logic.Cpr.Client.Sample/Program.cs]: changed as suggested

File: src/Kmd.Logic.Cpr.Client/CprClient.cs (hunk 25 above)
Comment (NakulSeerpu) [reply src/Kmd.Logic.Cpr.Client/CprClient.cs]: changed as suggested

File: README.md (hunk 26 above)
Comment (NakulSeerpu) [reply README.md:112]: changed as asuggested.

**Pull request merged**